### PR TITLE
added gogle auth and role-based acces

### PR DESCRIPTION
**Pull request by Yurii-Romaniv**
In this pull request I added role-based access control and google authorization-authentication (without registration because it will not be used). All auth process works on port 8080. react-app on port 3000 just redirects user to 8080 if during try to load data (through fetch()) server throws authorization page.

https://user-images.githubusercontent.com/70263963/236038452-27aa6001-fc9e-4c5b-9a87-c80f0c0330a0.mp4

### CODE REVIEW

File: frontend/public/index.html [hunk 1]
```diff
@@ -1,21 +1,21 @@
 <!DOCTYPE html>
```
Comment (justme1135) [inline frontend/public/index.html]: Revert

File: frontend/src/App.css [hunk 2]
```diff
@@ -1,38 +1,38 @@
 .App {
```
Comment (justme1135) [inline frontend/src/App.css]: Same here

File: frontend/src/TestEdit.js [hunk 3]
```diff
@@ -117,7 +118,7 @@ function TestEdit() {
         item.addedIds = Array.from(item.addedIds);
         item.deletedIds = Array.from(item.deletedIds);
 
-        await fetch('/tests/' + (item.test.id ? item.test.id : 'new'), {
+        await fetch('/teachersHome/tests/' + (item.test.id ? item.test.id : 'new'), {
```
Comment (justme1135) [inline frontend/src/TestEdit.js]: use ?? operator here

File: frontend/src/App.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ function App() {
             <Switch>
                 <Route path='/' exact={true} component={Home}/>
                 <Route path='/home' exact={true} component={TestList}/>
-                <Route path='/tests/:id' component={TestEdit}/>
+                <Route path='/teachersHome/tests/:id' component={TestEdit}/>
```
Comment (justme1135) [inline frontend/src/App.js]: naming should be like `/teachers-home`

File: frontend/src/TestEdit.js [hunk 5]
```diff
@@ -38,7 +39,7 @@ function TestEdit() {
     const {id} = useParams();
 
     const {error, isLoading} = useQuery('fullTests', () =>
-            fetch(`/tests/${id}`).then(res => res.json()),
+            fetch(`/teachersHome/tests/${id}`, {mode: "no-cors"}).then(checkAuth),
```
Comment (justme1135) [inline frontend/src/TestEdit.js]: same here

File: frontend/src/TestList.js [hunk 6]
```diff
@@ -16,8 +17,7 @@ export default function TestList() {
     const [tests, setTests] = useState(EmptyTest);
 
     const {error, isLoading} = useQuery('tests', () =>
-            fetch('/tests/first5').then(res =>
-                res.json()),
+            fetch('teachersHome/tests/first5', {mode: "no-cors"}).then(checkAuth),
```
Comment (justme1135) [inline frontend/src/TestList.js]: `/teachers-home/tests/first5`

File: frontend/src/TestList.js [hunk 7]
```diff
@@ -58,7 +58,7 @@ export default function TestList() {
                                 <td>
                                     <ButtonGroup>
                                         <Button size="sm" color="primary" tag={Link}
-                                                to={"/tests/" + test.id}>Edit</Button>
+                                                to={"/teachersHome/tests/" + test.id}>Edit</Button>
```
Comment (justme1135) [inline frontend/src/TestList.js]: `/teachers-home`

File: frontend/src/TestList.js [hunk 8]
```diff
@@ -68,7 +68,7 @@ export default function TestList() {
                     </tbody>
                 </Table>
                 <div className="float-right">
-                    <Button color="success" tag={Link} to="/tests/new">Add test</Button>
+                    <Button color="success" tag={Link} to="/teachersHome/tests/new">Add test</Button>
```
Comment (justme1135) [inline frontend/src/TestList.js]: +1

File: frontend/src/checkAuth.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+export function checkAuth(response) {
+    if (response.type === "opaque") {
+        window.location.href = "http://localhost:8080/login";
```
Comment (justme1135) [inline frontend/src/checkAuth.js]: URLs should not be hardcoded: use env variables or properties

File: frontend/src/index.css [hunk 10]
```diff
@@ -1,13 +1,13 @@
 body {
-  margin: 0;
-  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
+    margin: 0;
```
Comment (justme1135) [inline frontend/src/index.css]: revert

File: frontend/src/reportWebVitals.js [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
-  if (onPerfEntry && onPerfEntry instanceof Function) {
-    import('web-vitals').then(({ getCLS, getFID, getFCP, getLCP, getTTFB }) => {
-      getCLS(onPerfEntry);
-      getFID(onPerfEntry);
-      getFCP(onPerfEntry);
-      getLCP(onPerfEntry);
-      getTTFB(onPerfEntry);
-    });
-  }
+    if (onPerfEntry && onPerfEntry instanceof Function) {
```
Comment (justme1135) [inline frontend/src/reportWebVitals.js]: revert

File: src/main/java/com/example/nlearn/controllers/GroupController.java [hunk 12]
```diff
@@ -10,7 +10,7 @@
 import java.util.List;
 
 @RestController
-@RequestMapping("/groups")
+@RequestMapping("/teachersHome/groups")
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/controllers/GroupController.java]: naming

File: src/main/java/com/example/nlearn/controllers/TestController.java [hunk 13]
```diff
@@ -20,37 +21,36 @@
 import java.util.List;
 
 @RestController
-@RequestMapping("/tests")
+@RequestMapping("teachersHome/tests")
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/controllers/TestController.java]: `/teachers-home`

**Review by justme1135 — CHANGES_REQUESTED** [review]

File: frontend/src/TestList.js [hunk 14]
```diff
@@ -3,6 +3,7 @@ import {Button, ButtonGroup, Container, Table} from 'reactstrap';
 import AppNavbar from './AppNavbar';
 import {Link} from 'react-router-dom';
 import {useQuery} from "react-query";
+import {checkAuth} from "./checkAuth"
```
Comment (getmane) [inline frontend/src/TestList.js:6]: Install prettier for autoformatting

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("studentsHome")
```
Comment (getmane) [inline src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: naming

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+public class StudentsHomeController {
+
+    @Autowired
+    private TestService testService;
+
+    @GetMapping("/first5")
```
Comment (getmane) [inline src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: remove mapping route or rename it to be without a number in the name

File: src/main/java/com/example/nlearn/controllers/TestController.java [hunk 17]
```diff
@@ -20,37 +21,36 @@
...
 public class TestController {
 
     @Autowired
     private TestService testService;
 
     @GetMapping("/first5")
```
Comment (getmane) [inline src/main/java/com/example/nlearn/controllers/TestController.java]: same here

File: src/main/java/com/example/nlearn/services/CustomOAuth2UserService.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
+import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Service;
+import java.util.List;
+
+
```
Comment (getmane) [inline src/main/java/com/example/nlearn/services/CustomOAuth2UserService.java]: We don't leave two newlines after import statements (also above)

File: src/main/java/com/example/nlearn/services/TestService.java [hunk 19]
```diff
@@ -57,9 +61,13 @@ public ResponseEntity createTest(FullTest fullTest) {
     }
 
 
-    public ResponseEntity updateTest(Integer id, FullTest fullTest) {
+    public ResponseEntity updateTest(Integer testId, FullTest fullTest, Integer userId, Boolean isAdmin) {
+
+        if (!(isAdmin || testRepository.getById(testId).getAuthor().getId() == userId))
```
Comment (getmane) [inline src/main/java/com/example/nlearn/services/TestService.java]: You should move long conditions to separate methods for readability

File: src/main/java/com/example/nlearn/services/TestService.java [hunk 20]
```diff
@@ -91,9 +99,10 @@ public ResponseEntity updateTest(Integer id, FullTest fullTest) {
     }
 
 
-    public FullTest getTest(Integer id) {
-        List<Question> questions = questionService.findByTestId(id);
-        Test test = testRepository.getById(id);
+    public FullTest getTest(Integer testId, User author, Boolean isAdmin) {
+        Test test = testRepository.getTestById(testId);
+        if (!(isAdmin || test.getAuthor().getId() == author.getId())) throw new ResourceAccessException("");
```
Comment (getmane) [inline src/main/java/com/example/nlearn/services/TestService.java]: Long condition

File: src/main/resources/templates/create_test.html [hunk 21]
```diff
@@ -15,104 +18,102 @@
...
-            data: {group: this.value},
-            dataType: "text",
-            success: function (data) {
-                if (JSON.parse(data).groupExist) {
-                    groupInput.style.backgroundColor = "green";
-                } else {
-                    groupInput.style.backgroundColor = "red";
+    groupInput = document.getElementById("floatingInput3")
+    groupInput.addEventListener("keyup", function (e) {
+        if (/(([a-z]{1,})|([A-Z]{1,}))-[1-9]{2}/.test(this.value)) {
```
Comment (getmane) [inline src/main/resources/templates/create_test.html]: This regex is very cool but I don't intend to spend an hour trying to break it down. 
If it cannot be shortened, at least move it as a variable

File: src/main/resources/templates/create_user.html [hunk 22]
```diff
@@ -3,9 +3,12 @@
...
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet"
+          integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">
     <meta name="viewport" content="width=device-width, initial-scale=1">
-    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/js/bootstrap.bundle.min.js" integrity="sha384-ygbV9kiqUc6oa4msXn9868pTtWMgiQaeYH7/t7LECLbyPA2x65Kgf80OJFdroafW" crossorigin="anonymous"></script>
+    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/js/bootstrap.bundle.min.js"
+            integrity="sha384-ygbV9kiqUc6oa4msXn9868pTtWMgiQaeYH7/t7LECLbyPA2x65Kgf80OJFdroafW"
+            crossorigin="anonymous"></script>
     <meta charset='utf-8'>
     <meta http-equiv='X-UA-Compatible' content='IE=edge'>
     <title>get charapters of your name</title>
```
Comment (getmane) [inline src/main/resources/templates/create_user.html]: characters?

File: src/main/resources/templates/teachers_home.html [hunk 23]
```diff
@@ -14,78 +17,72 @@
...
-
-
+        <table>
+            <tr th:each="test: ${tests}">
+                <td th:text="${test.id}"/>
+                <td th:text="${test.name}"/>
+                <td th:text="${test.group.id}"/>
+            </tr>
+        </table>
+        <a href="/home/create_test">Add one</a>
```
Comment (getmane) [inline src/main/resources/templates/teachers_home.html]: stick to one case in routes (kebab case)

**Review by getmane — CHANGES_REQUESTED** [review]

File: frontend/public/index.html (hunk 1 above)
Comment (Yurii-Romaniv) [reply frontend/public/index.html]: reverted

File: frontend/src/App.css (hunk 2 above)
Comment (Yurii-Romaniv) [reply frontend/src/App.css]: reverted

File: frontend/src/TestEdit.js (hunk 3 above)
Comment (Yurii-Romaniv) [reply frontend/src/TestEdit.js]: done

File: frontend/src/App.js (hunk 4 above)
Comment (Yurii-Romaniv) [reply frontend/src/App.js]: done

File: frontend/src/TestList.js (hunk 6 above)
Comment (Yurii-Romaniv) [reply frontend/src/TestList.js]: done

File: frontend/src/TestList.js (hunk 7 above)
Comment (Yurii-Romaniv) [reply frontend/src/TestList.js]: done

File: frontend/src/TestList.js (hunk 8 above)
Comment (Yurii-Romaniv) [reply frontend/src/TestList.js]: done

File: frontend/src/checkAuth.js (hunk 9 above)
Comment (Yurii-Romaniv) [reply frontend/src/checkAuth.js]: done

File: frontend/src/index.css (hunk 10 above)
Comment (Yurii-Romaniv) [reply frontend/src/index.css]: reverted

File: frontend/src/reportWebVitals.js (hunk 11 above)
Comment (Yurii-Romaniv) [reply frontend/src/reportWebVitals.js]: reverted

File: src/main/java/com/example/nlearn/controllers/GroupController.java (hunk 12 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/GroupController.java]: done

File: src/main/java/com/example/nlearn/controllers/TestController.java (hunk 13 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/TestController.java]: done

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java (hunk 15 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: done

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java (hunk 16 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: done

File: src/main/java/com/example/nlearn/controllers/TestController.java (hunk 17 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/TestController.java]: done

File: src/main/java/com/example/nlearn/services/CustomOAuth2UserService.java (hunk 18 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/services/CustomOAuth2UserService.java]: done

File: src/main/java/com/example/nlearn/services/TestService.java (hunk 19 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/services/TestService.java]: done

File: src/main/java/com/example/nlearn/services/TestService.java (hunk 20 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/services/TestService.java]: done

File: src/main/resources/templates/create_test.html (hunk 21 above)
Comment (Yurii-Romaniv) [reply src/main/resources/templates/create_test.html]: done

File: src/main/resources/templates/create_user.html (hunk 22 above)
Comment (Yurii-Romaniv) [reply src/main/resources/templates/create_user.html]: done

File: src/main/resources/templates/teachers_home.html (hunk 23 above)
Comment (Yurii-Romaniv) [reply src/main/resources/templates/teachers_home.html]: done

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+
+@RestController
+@RequestMapping("students-home")
+public class StudentsHomeController {
+
+    @Autowired
+    private TestService testService;
+
+    @GetMapping("/firstFive")
```
Comment (getmane) [inline src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: yeah, you got it a bit wrong. 
Better remove the mapping at all

**Review by getmane — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/nlearn/controllers/StudentsHomeController.java (hunk 24 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/StudentsHomeController.java]: done

**Review by getmane — APPROVED** [review]

File: src/main/java/com/example/nlearn/controllers/TestController.java [hunk 25]
```diff
@@ -20,37 +21,36 @@
 import java.util.List;
 
 @RestController
-@RequestMapping("/tests")
+@RequestMapping("teachers-home/tests")
```
Comment (justme1135) [inline src/main/java/com/example/nlearn/controllers/TestController.java:24]: prepend /

File: src/main/resources/application.properties [hunk 26]
```diff
@@ -4,6 +4,8 @@ spring.datasource.username=springApp
 spring.datasource.password=12345
 spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
 #spring.jpa.show-sql: true
-
-
+spring.security.oauth2.client.registration.google.client-id=944892395889-f1ipq1mgfkh47nq70651u2nj5kn74rlj.apps.googleusercontent.com
+spring.security.oauth2.client.registration.google.client-secret=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
+spring.security.oauth2.client.registration.google.redirect-uri=http://localhost:8080/login/oauth2/code/google
```
Comment (justme1135) [inline src/main/resources/application.properties:9]: change to env variable
and remove from git history =)

**Review by justme1135 — CHANGES_REQUESTED** [review]

File: src/main/resources/application.properties (hunk 26 above)
Comment (Yurii-Romaniv) [reply src/main/resources/application.properties:9]: done

File: src/main/java/com/example/nlearn/controllers/TestController.java (hunk 25 above)
Comment (Yurii-Romaniv) [reply src/main/java/com/example/nlearn/controllers/TestController.java:24]: fixed

**Pull request merged**